[@types/jasmine/ts3.1] expect(ArrayLike<T>).toEqual() missing ArrayLike<ObjectContaining<T>> type for `expected` parameter

I tried using the `@types/jasmine` package and had problems.

It seems that the `ArrayLikeMatchers<T>#toEqual(expected, ..)` is missing a possible type for the `expected` parameter: `ArrayLike<ObjectContaining<Element>>`.


I have some test code like this.
```
const elements = service.locate(locator, context); // Returns Element[]
expect(elements).toEqual(
          [jasmine.objectContaining({tagName: 'DIV', id: 'find-this'})]);
```

I get the following compile error.
```
error TS2345: Argument of type 'ObjectContaining<{ tagName: string; id: string; }>[]' is not assignable to parameter of type 'Any | Spy | ArrayLike<Element> | ObjectContaining<ArrayLike<Element>> | ArrayContaining<Element>'.
  Type 'ObjectContaining<{ tagName: string; id: string; }>[]' is not assignable to type 'ArrayLike<Element>'.
    Index signatures are incompatible.
      Type 'ObjectContaining<{ tagName: string; id: string; }>' is missing the following properties from type 'Element': assignedSlot, attributes, classList, className, and 123 more.

45           [jasmine.objectContaining({tagName: 'DIV', id: 'find-this'})]);
```

This test was valid/passing before we updated the typings, so it seems that the appropriate type is missing.

Authors: @kolodny, @sheetalkamat

--- a/types/jasmine/index.d.ts
+++ b/types/jasmine/index.d.ts
@@ -179,7 +179,12 @@ declare function waitsFor(latchMethod: () => boolean, failureMessage?: string, t
 declare function waits(timeout?: number): void;
 
 declare namespace jasmine {
-    type Expected<T> = T | ObjectContaining<T> | Any | Spy;
+    type ExpectedRecursive<T> = T | ObjectContaining<T> | {
+        [K in keyof T]: ExpectedRecursive<T[K]> | Any;
+    };
+    type Expected<T> = T | ObjectContaining<T> | Any | Spy | {
+        [K in keyof T]: ExpectedRecursive<T[K]>;
+    };
     type SpyObjMethodNames<T = undefined> =
         T extends undefined ?
             (ReadonlyArray<string> | {[methodName: string]: any}) :

--- a/types/jasmine/jasmine-tests.ts
+++ b/types/jasmine/jasmine-tests.ts
@@ -870,6 +870,28 @@ describe("jasmine.objectContaining", () => {
         }));
     });
 
+    it("can be used in a nested object", () => {
+        interface nestedFooType {
+            nested: {
+                a: number;
+                b: number;
+                bar: string;
+            };
+        }
+
+        const nestedFoo: nestedFooType = {
+            nested: {
+                a: 1,
+                b: 2,
+                bar: 's',
+            },
+        };
+
+        expect(nestedFoo).toEqual({
+            nested: jasmine.objectContaining({b: 2})
+        });
+    });
+
     describe("when used with a spy", () => {
         it("is useful for comparing arguments", () => {
             const callback = jasmine.createSpy('callback');

--- a/types/jasmine/ts3.1/index.d.ts
+++ b/types/jasmine/ts3.1/index.d.ts
@@ -182,7 +182,12 @@ declare function waitsFor(latchMethod: () => boolean, failureMessage?: string, t
 declare function waits(timeout?: number): void;
 
 declare namespace jasmine {
-    type Expected<T> = T | ObjectContaining<T> | Any | Spy;
+    type ExpectedRecursive<T> = T | ObjectContaining<T> | {
+        [K in keyof T]: ExpectedRecursive<T[K]> | Any;
+    };
+    type Expected<T> = T | ObjectContaining<T> | Any | Spy | {
+        [K in keyof T]: ExpectedRecursive<T[K]>;
+    };
     type SpyObjMethodNames<T = undefined> =
         T extends undefined ?
             (ReadonlyArray<string> | {[methodName: string]: any}) :

--- a/types/jasmine/ts3.1/jasmine-tests.ts
+++ b/types/jasmine/ts3.1/jasmine-tests.ts
@@ -872,6 +872,28 @@ describe("jasmine.objectContaining", () => {
         }));
     });
 
+    it("can be used in a nested object", () => {
+        interface nestedFooType {
+            nested: {
+                a: number;
+                b: number;
+                bar: string;
+            };
+        }
+
+        const nestedFoo: nestedFooType = {
+            nested: {
+                a: 1,
+                b: 2,
+                bar: 's',
+            },
+        };
+
+        expect(nestedFoo).toEqual({
+            nested: jasmine.objectContaining({b: 2})
+        });
+    });
+
     describe("when used with a spy", () => {
         it("is useful for comparing arguments", () => {
             const callback = jasmine.createSpy('callback');

